kong/kong.lua

```diff
@@ -157,6 +161,34 @@ function Kong.ssl_certificate()
   end
 end
 
+function Kong.balancer()
+  local addr = ngx.ctx.balancer_address
+  addr.tries = addr.tries + 1
+  if addr.tries > 1 then  
+    -- only call balancer on retry, first one is done in `core.access.before` which runs
+    -- in the ACCESS context and hence has less limitations than this BALANCER context
+    -- where the retries are executed
+    
+    -- record failure data
+    addr.failures = addr.failures or {}
+    local state, code = ngx_balancer.get_last_failure()
+    addr.failures[addr.tries-1] = { name = state, code = code }
```

Comment:
We can use the [`set_more_tries` function](https://github.com/openresty/lua-resty-core/blob/master/lib/ngx/balancer.md#set_more_tries) on the first try. Using a sensible default (5 ??) and make it configurable per api. Regarding logging; nginx will do the actual retry check and block if it hits it. So no extra logging required?